### ISSUE TITLE
[Announcement]: master branch will move to support Spark 3.0.

### ISSUE THREAD

**Issue by imback82** [microsoft/hyperspace#407]
The `master` branch is planned to move to have a dependency on Spark 3.0 after Hyperspace 0.5 is released; this indicates that Spark 2.4 will be supported by Hyperspace version 0.5.x. `branch-0.5` will be created at that time, and any critical bugs will be back-ported and new patch versions (e.g., 0.5.1, 0.5.2) will be released if necessary. However, there is no plan to back-port any new features.

Please leave comments if you have any concerns.

**Comment by rapoth** [comment]
CC: @andrei-ionescu

**Comment by andrei-ionescu** [comment]
I'll just layout my thoughts on it...

1. We need to make sure that the Spark APIs that Hyperspace uses is still available in Spark 3 and works as expected. I know the Catalyst APIs - even though the classes are `public` classes - are considered as "experimental" and not public as "public APIs" which are not supported by all Spark compute providers/vendors.
2. It will be needed to clearly state and explain what gets removed from main branch once the switch to Spark 3 happens:
    1. Drop support for Scala 2.11.
    2. Drop support for Spark 2.x.
    3. Any changes to Python version?
    4. Any changes into the supported JVM?
3. It will need to have a guiding "cut-off date". I say "guiding" as that date may change as thing progress. For example: _"30th of September will be the day we target to switch to Spark 3 on main branch."_
4. Major features will need to be completed on the current version and then brought into Spark 3.
    1. We need a list with what feature is started and partially committed (ie: the nested field support).
    2. We need a list with what is not started but must be completed before the cut-off date.
    3. Need to track the progress on these mandatory features, tasks, etc.
5. Hyperspace should still be on 2.4 for at least up to the end of the year as the major players with Spark based compute services are still supporting 2.4. Removing the 2.x support too early will hinder the growth of community around Hyperspace as from what I'm aware the majority is still on Spark 2.x.
    1. Azure Synapse is 2.4 (correct me if I'm wrong) and I don't know when it will switch to 3.
    2. Cloudera is still on Spark 2 and no signs to switch to 3.
    2. Databricks is supporting Spark 2 up to the end of the year then dropping the support for Spark 2.x altogether.
    3. Qubole is on Spark 2 and no signs to switch to 3.
    4. Can we have a better plane view on what's the adoption of Spark 3 in the market?
6. Regarding "not planning to back port any new features" I would let that to the contributor/collaborator/person that is interested into that feature. If that person wants and really needs that feature in 2.4 the person should invest the effort to do the work. Then the community will review it but with a lower priority than the PR in the main branch. A new feature should be added first to the master and if necessary and possible that can be back ported into 2.4 by the one interested in that new feature.
7. Why can't we go with both Spark 2.x and 3.x in parallel? I think we can keep the majority of the repo as it is we need just to extract in separate modules the different implementations for Spark 2 and 3. Can we build the following?
    1. `hyperspace_2.11: 0.6.0`
    2. `hyperspace_2.12: 0.6.0`
    3. `hyperspace_spark3_2.12: 0.6.0`

**Comment by imback82** [comment]
Thanks @andrei-ionescu for the feedback.

So, it comes down to two: 1) backport new features to 0.6 to support Spark 2.4, or 2) different JARs each targeting specific Spark. The reason we chose 1) was the upfront cost for supporting 2). However, I think 2) will pay the dividends in the long term; for example, if 3.1 is not supported, we have to cut another branch and support.

I will update this thread once we do some initial costing to support 2). Please let me know if you are also interested in this work. Thanks.

**Comment by andrei-ionescu** [comment]
@imback82, I'm interested in this work but my focus at this moment is on completing the "Support nested fields" feature. It seems that it will take more time than I expected. It would be helpful for me to understand your timeframe (or the roadmap deadlines) on delivering the Spark 3 support.

**Comment by clee704** [comment]
If the changes between Spark 2 and 3 are not significant in terms of API, public or internal, then we can go with the option 7, producing JAR for each Spark version. I'll take a look if this is feasible.

**Comment by imback82** [comment]
We will go with option 7 thanks to @clee704 (#421). (I think we need to change the JAR names to conform to Maven spec, but I will check and take care of them).
